### ISSUE TITLE
IDE analyzer: suggest null-conditional assignment (C# 14) to simplify null checks

### ISSUE THREAD

**Issue by glen-84** [dotnet/roslyn#83195]
## Summary

C# 14 introduced null-conditional assignment on the left-hand side of an assignment ([docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-)). There is an opportunity for a new IDE diagnostic (similar to `IDE0031` which suggests null propagation for reads) that suggests simplifying null-guard-then-assign patterns.

### Before

```csharp
if (obj is not null)
{
    obj.Property = value;
}
```

### After

```csharp
obj?.Property = value;
```

This also applies to indexer access:

```csharp
// Before
if (list is not null)
{
    list[0] = value;
}

// After
list?[0] = value;
```

## Patterns to detect

The analyzer should detect `if` statements whose sole body is a single assignment guarded by a null check on the receiver, including variations like:

- `if (x != null)` / `if (x is not null)` / `if (x is { })` (and negated forms with early return)
- Member access (`x.Prop = ...`) and indexer access (`x[i] = ...`)

## Semantic considerations

The right-hand side is only evaluated when the left-hand side is non-null. The transformation is safe when skipping the RHS evaluation is acceptable (e.g., when the RHS is a literal, local variable, parameter, or other side-effect-free expression). The analyzer should account for this.

## Precedent

- `IDE0031` — Use null propagation (for reads)
- `IDE0270` — Use null coalescing
- `IDE0029` / `IDE0030` — Use coalesce expression

---

*This issue was authored with the help of Claude Code (AI assistant).*

**Comment by CyrusNajmabadi** [comment]
This analyzer already exists.

**Comment by glen-84** [comment]
Sorry, but I can't find it – what is the diagnostic ID?

**Comment by CyrusNajmabadi** [comment]
<img width="657" height="356" alt="Image" src="https://github.com/user-attachments/assets/11cc62a5-eda7-4c7f-ba18-3f1305fec5fc" />

IDE0031

**Comment by RikkiGibson** [comment]
If you are not seeing these kinds of IDE0031 suggestions reported, then your tooling might be out of date.

**Comment by glen-84** [comment]
Ah, my apologies, I had suppressed `IDE0031` while using Roslynator's `RCS1146` instead for other types of conditional access.

The ReSharper inspection is able to catch some other cases (not sure if this is covered by another IDE rule yet, as I haven't set them all to enabled):

```cs
if (s == null)
{
    return null;
}

return s.Replace("\\", "\\\\") // *** Could use `s?.` here instead of the above check.
    .Replace("\"", "\\\"")
    .Replace("\n", "\\n")
    .Replace("\r", "\\r")
    .Replace("\t", "\\t");
```

(and the inverse [s != null])

```cs
if (additionalMockSetup is not null)
{
    additionalMockSetup(mock);
}

// -> additionalMockSetup?.Invoke(mock);
```

These ones are also flagged by ReSharper, but they're quite complex.

```cs
var parameter = node.AncestorsAndSelf().OfType<ParameterSyntax>().FirstOrDefault();
if (parameter is null) // *** Here
{
    return;
}

// Find the method declaration
var methodDeclaration = parameter.AncestorsAndSelf().OfType<MethodDeclarationSyntax>().FirstOrDefault();
if (methodDeclaration is null)
{
    return;
}
```

```cs
var attribute = node.AncestorsAndSelf().OfType<AttributeSyntax>().FirstOrDefault();
if (attribute is null) // *** Here
{
    return;
}

// Find the attribute list
var attributeList = attribute.Parent as AttributeListSyntax;
if (attributeList is null)
{
    return;
}
```

```cs
if (_appServices is not null) // *** Here
{
    var service = _appServices.GetService<T?>();

    if (service is not null)
    {
        return service;
    }
}
```

This one has a [bug](https://youtrack.jetbrains.com/issue/RSRP-503516/Use-null-propagation-quick-fix-on-cast-expression-omits-nullable-annotation-on-target-type) in ReSharper:

```cs
if (_dict is not null)
{
    ((ICollection<KeyValuePair<string, object?>>)_dict).CopyTo(array, arrayIndex);
}
```

**Comment by S-Luiten** [comment]
This should be a separate analyzer, or it should be possible to configure IDE0031 to not report on this new syntax. This is especially annoying when you also configured IDE0058 to require a discard, see https://github.com/dotnet/roslyn/issues/80766

**Comment by CyrusNajmabadi** [comment]
> This is especially annoying when you also configured IDE0058 to require a discard

The bug seems to be in IDE0058.  It should not require a discard on this new syntax.
